File: 2024/src/day2/main.clj
(ns day2.main
  (:require [clojure.string :as str]))

(defn parse-lines [lines]
  (for [line lines]
    (as-> line v
          (str/split v #"\s+")
          (mapv ^[String] Integer/parseInt v))))

(defn parse-text [text]
  (-> text
      (str/split-lines)
      (parse-lines)))

(defn parse-file [file]
  (-> file
      (slurp)
      (parse-text)))

; part 1

(defn is-safe [row]
  (and
    (or
      (every? #(> 0 %) row)
      (every? #(< 0 %) row))
    (every? #(<= 1 (abs %) 3) row)))

(defn part1-row [row]
  (->> row
       (partition 2 1)
       (map #(apply - %))
       (is-safe)))

(defn part1 [input]
  (->> input
       (map part1-row)
       (remove false?)
       (count)
       (println)))

(part1 (parse-text "7 6 4 2 1\n1 2 7 8 9\n9 7 6 2 1\n1 3 2 4 5\n8 6 4 4 1\n1 3 6 7 9"))

(part1 (parse-file "src/day2/input.txt"))

; part 2

(defn part2-row [row]
  (as-> row v
       (map
         #(into (subvec v 0 %) (subvec v (inc %)))
         (range (count v)))
       (map part1-row v)
       (some identity v)))



(defn part2 [input]
  (->> input
       (map part2-row)
       (remove nil?)
       (count)
       (println)))

(part2 (parse-text "7 6 4 2 1\n1 2 7 8 9\n9 7 6 2 1\n1 3 2 4 5\n8 6 4 4 1\n1 3 6 7 9"))

(part2 (parse-file "src/day2/input.txt"))

File: 2024/src/day2/main.py
# part 1

print(
    sum(
        (all(x > 0 for x in d) or all(x < 0 for x in d))
        and all(1 <= abs(x) <= 3 for x in d)
        for l in open(__file__.replace("\\", "/").rsplit("/", 1)[0] + "/input.txt")
        .read()
        .splitlines()
        for r in [[int(n) for n in l.split()]]
        for d in [[a - b for a, b in zip(r, r[1:])]]
    )
)

# part 2

print(
    sum(
        any(
            (all(x > 0 for x in d) or all(x < 0 for x in d))
            and all(1 <= abs(x) <= 3 for x in d)
            for p in [r[:i] + r[i + 1 :] for i in range(len(r))]
            for d in [[a - b for a, b in zip(p, p[1:])]]
        )
        for l in open(__file__.replace("\\", "/").rsplit("/", 1)[0] + "/input.txt")
        .read()
        .splitlines()
        for r in [[int(n) for n in l.split()]]
    )
)

# lmao.

# fmt: off
print(sum(any((all(x > 0 for x in d) or all(x < 0 for x in d)) and all(1 <= abs(x) <= 3 for x in d) for p in [r[:i] + r[i + 1 :] for i in range(len(r))] for d in [[a - b for a, b in zip(p, p[1:])]]) for l in open(__file__.replace("\\", "/").rsplit("/", 1)[0] + "/input.txt").read().splitlines() for r in [[int(n) for n in l.split()]]))
# fmt: on
